ENH: Generalized LED control using Beckhoff Voltage/Current Output raw value scaling

Mbosum opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
## Description
<!--- Describe your changes in detail -->
Developed a function block for 24V PWM LED control. However, this FB is general enough that it could be used with just a straight analog terminal. It should be noted, although, that the raw bits that are sent to the terminal output will be 50% of max or 2^14 in order to keep the duty cycle below 50%.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
We have a 12 V max LED circuit on vacuum chamber viewport illuminators, which showed a bad dimming response curve to simple voltage modulation via the EL4004 AO terminal. Also, this terminal only goes up to 10 V. 

We use an EL2502 24 V PWM output terminal instead, and set the maximum duty cycle to 50% in the FB so the max effective voltage is 12 V. Furthermore, we demonstrated a much better dimming response curve via PWM.

<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

I have been testing the FB in a test twincat solution and test IOC:
https://github.com/rsmm97/LED_test
/reg/g/pcds/epics-dev/bosum123/plc_iocs/LED_test

## Where Has This Been Documented?
<!--  Include where the changes made have been documented. -->
<!--  This can simply be  a comment in the code or updating a docstring -->

https://confluence.slac.stanford.edu/display/L2SI/IP1+LED+controls (Work in progress)
## Pre-merge checklist
- [x] Code works interactively
- [x] Code contains descriptive comments
- [x] Test suite passes locally
- [ ] Libraries are set to ``Always Newest`` version (``Library, *``)
- [ ] Committed with ``pre-commit`` or ran ``pre-commit run --all-files``

klauer commented on LCLSGeneral/LCLSGeneral/POUs/Hardware/FB_LED_PWM.TcPOU:
```diff
@@ -0,0 +1,56 @@
+
+    {attribute 'pytmc' := '
+        pv: CIL:PCT
+        EGU: %
```
```suggestion
        field: EGU %
```

klauer replied:
And to clarify, this is a value of 0 - 100 and not normalized 0 - 1.0, correct?

klauer commented on LCLSGeneral/LCLSGeneral/POUs/Hardware/FB_LED_PWM.TcPOU:
```diff
@@ -0,0 +1,56 @@
+END_IF]]></ST>
+    </Implementation>
+    <Method Name="M_Set_IllPercentage" Id="{dbd69c00-c8e8-4492-8041-fbc83238a849}">
+      <Declaration><![CDATA[METHOD M_Set_IllPercentage : LREAL
+VAR_INPUT
+	IlluminatorPercent : LREAL;
+END_VAR
+]]></Declaration>
+      <Implementation>
+        <ST><![CDATA[This^.fIlluminatorPercent := IlluminatorPercent;]]></ST>
+      </Implementation>
+    </Method>
+  </POU>
```
Why have a method for this? Shouldn't `fIlluminatorPercent` be `VAR_INPUT`?

Mbosum replied:
That's a good idea, let me try that. This was my first iteration to "get things to work" so to say, so I saw that we had previously been setting variables via methods in the vacuum library.

klauer commented on LCLSGeneral/LCLSGeneral/POUs/Hardware/FB_LED_PWM.TcPOU:
```diff
@@ -0,0 +1,56 @@
+    fReal:=fIlluminatorPercent,
+    fSafeMax:=100,
+    fSafeMin:=0,
+    iTermBits:=14, // want to set scaling to 50% duty cycle = maximum raw bits sent to terminal output
```
This block is a pretty thin wrapper around `FB_AnalogOutput`; should it be providing control over `iTermBits`?

Mbosum replied:
I think so, because iTermBits determines the max raw value that can be sent to the output. In this case I want the max value (50% duty cycle) to be only half of what the nominal output can be.

ZLLentz replied:
I think the question here is more: for generalizing this FB, shouldn't this be an input variable rather than a hard-coded 14?

Mbosum replied:
Yeah I agree, let me see if I can get it to work with the raw bits being an input variable, set at the project level.

klauer reviewed (commented):
I'm not 100% sure a customized wrapper for LED control fits in lcls-twincat-general, but I could be convinced. 

There are a few things to fix up in the implementation, too, but thinking about how generally useful this could be:
* Do you expect a lot of projects to use this block? 
* Will they all use the same output terminal?
* Is there a scenario where you'd want to control the maximum/minimum illumination?
* Might the definition of "LED is powered" depend on a configurable threshold? I'd guess a 1% duty cycle would not be visible

Mbosum commented:
> I'm not 100% sure a customized wrapper for LED control fits in lcls-twincat-general, but I could be convinced.
> 
> There are a few things to fix up in the implementation, too, but thinking about how generally useful this could be:
> 
> * Do you expect a lot of projects to use this block?
> * Will they all use the same output terminal?
> * Is there a scenario where you'd want to control the maximum/minimum illumination?
> * Might the definition of "LED is powered" depend on a configurable threshold? I'd guess a 1% duty cycle would not be visible

1. At the very least I see 3 projects using this (every endstation in TMO). RIX may also follow suit which would add more.
2. All the projects in TMO will use the EL2502 output terminal, as I have not found a better alternative so far. RIX I'm not sure yet.
3. Yes, it would in fact be a nice feature to be able to control the max/min illumination.
4. Interestingly, even this PWM control of these LED rings has not shown a linear dimming response curve. 50 % duty corresponds to full illumination, but I could only get them to about 40% illumination with 1% duty. The most essential capability here is definitely remote on/off control, with dimming being a nice add on. I was unable to get a circuit diagram or datasheet from the vendor so I was kind of left guessing about the causes behind this nonlinearity. I think the bLedPower variable is good just as a sanity check; is it completely OFF or getting some kind of power?

ZLLentz commented:
I think `lcls-twincat-general` is a good choice for function blocks for general-purpose device controls. LEDs may be added to all sorts of PLC systems.

I think function blocks that are tuned for very specific cases are also :+1:, they just need to be named and documented accordingly.

ZLLentz commented on LCLSGeneral/LCLSGeneral/POUs/Hardware/FB_LED_PWM.TcPOU:
```diff
@@ -0,0 +1,56 @@
+    Reads an percentage of illumination determined by the user and converts it to a raw value
+    to send to the terminal output via the FB_AnalogOutput.
+    2022-5-03 Maarten Thomas-Bosum
+*)
```
I think some of the PR text should make it into the documentation comment here- assumptions made, devices used on, etc.

ZLLentz commented on LCLSGeneral/LCLSGeneral/POUs/Hardware/FB_LED_PWM.TcPOU:
```diff
@@ -0,0 +1,56 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<TcPlcObject Version="1.1.0.1" ProductVersion="3.1.4022.18">
+  <POU Name="FB_LED_PWM" Id="{5fae4b24-b1d6-4ac5-a010-9aff4e2893db}" SpecialFunc="None">
```
I think we should take a moment to pause and consider if this is a generalized LED_PWM FB or if it's a very specific ELsomethingsomething FB. I think matching the identity is important.

ZLLentz commented:
I'm overall :+1: on this btw!

klauer commented:
> I think `lcls-twincat-general` is a good choice for function blocks for general-purpose device controls. LEDs may be added to all sorts of PLC systems.
> 
> I think function blocks that are tuned for very specific cases are also 👍, they just need to be named and documented accordingly.

This is exactly what I was after, but phrased much better. A few more fixes and this should be good @Mbosum 👍

Mbosum commented:
I renamed the FB to FB_LED because I changed the iTermBits variable to a VAR_INPUT, so it would be configurable according to the setup. That being said, it is still bounded by the inherent Beckhoff full scale value of (2^15  - 1) = 32767.

I also updated the FB docstring to give detailed information about this and use cases.

Completed the couple other small fixes as well.

ZLLentz commented on LCLSGeneral/LCLSGeneral/POUs/Hardware/FB_LED.TcPOU:
```diff
@@ -0,0 +1,62 @@
+VAR
+	iIlluminatorINT AT %Q*: INT;
+
+    {attribute 'pytmc' := '
+        pv: PWR
+        field: ZNAM OFF
+        field: ONAM ON
+    '}
+    bLedPower AT %Q*: BOOL;
+
+    {attribute 'pytmc' := '
+        pv: CIL:PCT
+        field: EGU %
+    '}
+    fIlluminatorPercent: LREAL;
```
One last small thing to consider: PV names
PWR = power, makes sense
CIL:PCT = camera illuminator, colon, percent
Is the CIL part intended or left over from a copy/paste?

Mbosum replied:
I think it is intended. Other LED illuminators have the CIL fungible term included in their PV names.

ZLLentz replied:
Okay, if you intend for this FB to only be used for cameras, then that's ok. Just make sure to add that to the docs.

Mbosum replied:
I changed this to "ILL:PCT" now

klauer commented on LCLSGeneral/LCLSGeneral/POUs/Hardware/FB_LED.TcPOU:
```diff
@@ -0,0 +1,62 @@
+    2022-5-03 Maarten Thomas-Bosum
+*)
+VAR_INPUT
+	iTermBits : UINT;
```
`fIlluminatorPercent` should be an input, should it not?

Mbosum replied:
I tested the code as is by creating a test IOC, and I was able to caget and caput the illuminator percentage fine from EPICS. 

I can try to make fIlluminatorPercent an input variable and see how it functions.

klauer replied:
As an input, you have the option of controlling it from the PLC or from EPICS.

It seems unnecessarily restrictive to say "this is meant only for EPICS".

Thoughts on the PV name there as @ZLLentz had pointed out [above](https://github.com/pcdshub/lcls-twincat-general/pull/63#discussion_r871547516), @Mbosum?

Mbosum replied:
@ZLLentz I'm not convinced we need bLedPower to be controllable from EPICS. I think I at least initially just wanted it to be there as a sanity check when the user sets the illumination percentage.

ZLLentz replied:
> I'm not convinced we  we need bLedPower to be controllable from EPICS

As written, it's _only_ controllable from EPICS.

Mbosum replied:
I changed it to an input variable now

Mbosum commented:
Ken I suggest changing the title, since this is now usable for applications outside of just El2502 PWM control. The Beckhoff analog voltage and current control modules all scale from 0 to 32767 FSV, hence having the iTermBits configurable makes this FB applicable to a wide range of controls.

klauer commented:
Thanks for working with us on this, @Mbosum! Sorry for the long process - I think we could have sat down and worked through all this in person in like a half hour, but it ended up getting dragged out. I think this should be good-to-go. 👍
